spirv_msl.cpp

```diff
@@ -166,11 +167,25 @@ void CompilerMSL::build_implicit_builtins()
 		uint32_t workgroup_id_type = 0;
 
 		ir.for_each_typed_id<SPIRVariable>([&](uint32_t, SPIRVariable &var) {
-			if (var.storage != StorageClassInput || !ir.meta[var.self].decoration.builtin)
+			if (!ir.meta[var.self].decoration.builtin)
 				return;
 
 			// Use Metal's native frame-buffer fetch API for subpass inputs.
 			BuiltIn builtin = ir.meta[var.self].decoration.builtin_type;
+
+			if (var.storage == StorageClassOutput)
+			{
+				if (need_sample_mask && builtin == BuiltInSampleMask)
+				{
+					builtin_sample_mask_id = var.self;
+					mark_implicit_builtin(StorageClassOutput, BuiltInSampleMask, var.self);
+					does_shader_write_sample_mask = true;
```

Comment:
So I suppose this will be set to true if shader merely declares the SampleMask variable (even if it's unused)?

Reply:
Yes. I'm not sure what's the behaviour that we want in this case. If declaring the SampleMask as an output and not writing to it anything means that the output will be 0, then we need to assume that the shader writes to the sample mask.

Reply:
Perhaps emit an input fixup to force sample mask to 0xffffffff as well?

Reply:
In that case, the variable should be called does_shader_declare_sample_mask perhaps to avoid confusion.

Reply:
Emitting an input fixup might change the current behaviour of shaders if what I mentioned is true and declaring but not explicitly writing to the sample_mask means writing 0 to it. I don't know if this is the desired behaviour though.

Reply:
I think you can look at the entry point declaration to see if the SampleMask is used.

```
A variable in the Output storage class decorated with SampleMask is an array of integers forming a
bit array in a manner similar an input variable decorated with SampleMask, but where each bit
represents coverage as computed by the shader. Modifying the sample mask by writing zero to a
Chapter 14. Shader Interfaces | 795bit of SampleMask causes the sample to be considered uncovered. If this variable is also decorated
with OverrideCoverageNV, the fragment coverage is replaced with the sample mask bits set in the
shader otherwise the fragment coverage is ANDed with the bits of the sample mask. If the
fragment shader is being evaluated at any frequency other than per-fragment, bits of the sample
mask not corresponding to the current fragment shader invocation are ignored. This array must
be sized in the fragment shader either implicitly or explicitly, to be no larger than the
implementation-dependent maximum sample-mask (as an array of 32-bit elements), determined
by the maximum number of samples. If a fragment shader entry point’s interface includes an
output variable decorated with SampleMask, the sample mask will be undefined for any array
elements of any fragment shader invocations that fail to assign a value. If a fragment shader
entry point’s interface does not include an output variable decorated with SampleMask, the
sample mask has no effect on the processing of a fragment.
```

Reply:
So, an input fixup is not necessary. The application must assign something to it if it exists in the entry point. If it does not exist in the entry point, then we need to declare our own sample mask.

Reply:
Thanks, I think the current code is fine in this case.

Reply:
I don't see any check for the entry point interface variables though, this code will iterate over all variables in the module if I understand it correctly?

Reply:
It seems so, it's doing the same thing as for the other builtin variables. There are no entry point interface checks for BuiltInSampleID either.

Reply:
Hm, ok. Please just add a FIXME then so we don't forget that this might be something to revisit later.

Reply:
Added!